core/src/main/scala/io/delta/sql/DeltaSparkSessionExtension.scala

```diff
@@ -91,7 +92,7 @@ class DeltaSparkSessionExtension extends (SparkSessionExtensions => Unit) {
       new PreprocessTableDelete(session.sessionState.conf)
     }
     extensions.injectOptimizerRule { session =>
-      new ActiveOptimisticTransactionRule(session)
+      new PrepareDeltaScan(session)
```

Comment:
In the description you say that  this change enables "query reuse that snapshot". But isnt that what ActiveOptimisticTxnRule already doing? Whats different with this approach?

Reply:
The current code only works with existing transactions (e.g. `if ... && OptimisticTransaction.getActive().nonEmpty`) whereas this new code is coupled to the lifetime of the rule. so in addition to the current txn it will save snapshots for the actual file index + delta log. it does this by saving them to a concurrent hash map, to avoid unnecessary recomputation.